File: core/src/com/exp4j/Helpers/EquationHelper.java
package com.exp4j.Helpers;

import java.util.HashMap;
import java.util.Random;

import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.utils.ObjectFloatMap;
import com.badlogic.gdx.utils.ObjectIntMap;
import com.exp4j.Functions.ChanceFunction;
import com.exp4j.Functions.MathUtilFunctions;
import com.exp4j.Functions.RandomFunction;
import com.exp4j.Operators.BooleanOperators;
import com.lyeeedar.Enums;
import net.objecthunter.exp4j.Expression;
import net.objecthunter.exp4j.ExpressionBuilder;

public class EquationHelper
{
	private static final ObjectFloatMap<String> emptyMap = Enums.Statistic.getEmptyMap();

	public static Expression tryBuild( ExpressionBuilder expB )
	{
		Expression exp = null;

		// try
		// {
		exp = expB.build();
		// }
		// catch (Exception e) { }

		return exp;
	}

	public static void setVariableNames( ExpressionBuilder expB, ObjectFloatMap<String> variableMap, String prefix )
	{
		for ( String key : variableMap.keys() )
		{
			expB.variable( prefix + key );
		}
	}

	public static void setVariableValues( Expression exp, ObjectFloatMap<String> variableMap, String prefix )
	{
		for ( String key : variableMap.keys() )
		{
			exp.setVariable( prefix + key, variableMap.get( key, 0f ) );
		}
	}

	public static ExpressionBuilder createEquationBuilder( String eqn )
	{
		return createEquationBuilder( eqn, MathUtils.random );
	}

	public static ExpressionBuilder createEquationBuilder( String eqn, Random ran )
	{
		ExpressionBuilder expB = new ExpressionBuilder( eqn );
		BooleanOperators.applyOperators( expB );
		expB.function( new RandomFunction( ran ) );
		expB.function( new ChanceFunction( ran ) );
		MathUtilFunctions.applyFunctions( expB );

		return expB;
	}

	public static boolean isNumber( String string )
	{
		if ( string == null || string.isEmpty() ) { return false; }
		int i = 0;
		if ( string.charAt( 0 ) == '-' )
		{
			if ( string.length() > 1 )
			{
				i++;
			}
			else
			{
				return false;
			}
		}
		for ( ; i < string.length(); i++ )
		{
			if ( !Character.isDigit( string.charAt( i ) ) ) { return false; }
		}
		return true;
	}

	public static float evaluate( String eqn )
	{
		return evaluate( eqn, emptyMap, MathUtils.random );
	}

	public static float evaluate( String eqn, Random ran )
	{
		return evaluate( eqn, emptyMap, ran );
	}

	public static float evaluate( String eqn, ObjectFloatMap<String> variableMap )
	{
		return evaluate( eqn, variableMap, MathUtils.random );
	}

	public static float evaluate( String eqn, ObjectFloatMap<String> variableMap, Random ran )
	{
		if ( isNumber( eqn ) )
		{
			return Integer.parseInt( eqn );
		}
		else
		{
			ExpressionBuilder expB = createEquationBuilder( eqn, ran );
			setVariableNames( expB, variableMap, "" );
			Expression exp = tryBuild( expB );

			if ( exp == null )
			{
				return 0;
			}
			else
			{
				setVariableValues( exp, variableMap, "" );
				double rawVal = exp.evaluate();

				return (float)rawVal;
			}
		}
	}
}

File: core/src/com/exp4j/Helpers/EquationHelper.kt
package com.exp4j.Helpers

import com.badlogic.gdx.utils.ObjectFloatMap
import com.exp4j.Functions.ChanceFunction
import com.exp4j.Functions.MathUtilFunctions
import com.exp4j.Functions.RandomFunction
import com.exp4j.Operators.BooleanOperators
import com.exp4j.Operators.PercentageOperator
import com.lyeeedar.Util.Random
import net.objecthunter.exp4j.Expression
import net.objecthunter.exp4j.ExpressionBuilder

class EquationHelper
{
	companion object
	{
		fun setVariableNames(expB: ExpressionBuilder, variableMap: ObjectFloatMap<String>)
		{
			for (key in variableMap.keys())
			{
				expB.variable(key)
			}
		}

		fun setVariableValues(exp: Expression, variableMap: ObjectFloatMap<String>)
		{
			val valuesToBeSet = exp.variableNames

			for (key in variableMap.keys())
			{
				exp.setVariable(key, variableMap.get(key, 0f).toDouble())

				valuesToBeSet.remove(key)
			}

			for (variable in valuesToBeSet)
			{
				exp.setVariable(variable, 0.0)
			}
		}

		fun evaluate(eqn: String, variableMap: ObjectFloatMap<String> = ObjectFloatMap(), seed: Long? = null): Float
		{
			try
			{
				return eqn.toFloat()
			}
			catch (ex: Exception)
			{
				val expB = ExpressionBuilder(eqn)
				expB.exceptionOnMissingVariables = false

				BooleanOperators.applyOperators(expB)
				expB.operator(PercentageOperator.operator)

				val seed = seed ?: Random.random.nextLong()
				val randomFun = RandomFunction.obtain()
				val chanceFun = ChanceFunction.obtain()
				randomFun.set(seed)
				chanceFun.set(seed)

				expB.function(randomFun)
				expB.function(chanceFun)
				MathUtilFunctions.applyFunctions(expB)

				setVariableNames(expB, variableMap)
				val exp = expB.build()

				val expectedVariables = exp.variableNames
				for (name in expectedVariables)
				{
					if (!variableMap.containsKey(name))
					{
						exp.setVariable(name, 0.0)
					}
				}

				if (exp == null)
				{
					randomFun.free()
					chanceFun.free()
					return 0f
				}
				else
				{
					setVariableValues(exp, variableMap)
					val rawVal = exp.evaluate()

					randomFun.free()
					chanceFun.free()

					return rawVal.toFloat()
				}
			}
		}
	}
}

fun String.unescapeCharacters(): String
{
	var output = this
	output = output.replace("&gt;", ">")
	output = output.replace("&lt;", "<")
	output = output.replace("&amp;", "&")
	return output
}

fun String.evaluate(variableMap: ObjectFloatMap<String> = ObjectFloatMap(), seed: Long? = null): Float = EquationHelper.evaluate(this, variableMap, seed)